driver/src/orderbook/streamed/state.rs

```diff
@@ -11,47 +11,75 @@ use std::collections::HashMap;
 use std::iter::Iterator;
 use thiserror::Error;
 
-// TODO: Should we handle https://github.com/gnosis/dex-contracts/issues/620 ?
-// There is a workaround detailed in the issue it hasn't been implemented.
-
 // Most types, fields, functions in this module mirror the smart contract because we need to
 // emulate what it does based on the events it emits.
+//
+// We keep track of the pending solution for the batch separately instead of applying the trades
+// immediately. This ensures that orders and balances have correctly *not* been updated by the
+// pending solution until the solution becomes the real solution.
+// This has the side effect of simplifying solution revert logic by allowing us to replace the whole
+// pending solution instead of having to revert individual trades.
 
 /// The orderbook state as built from received events.
 ///
 /// Note that there is no way to revert an event. The order in which events are received matters
 /// Applying events `A B` does not always result in the same state as `B A`. For example, a Trade
 /// can only be processed if the OrderPlacement for the traded order has previously been observed.
-#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct State {
     orders: HashMap<(UserId, OrderId), Order>,
     balances: HashMap<(UserId, TokenAddress), Balance>,
     tokens: Tokens,
+    pending_solution: PendingSolution,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            orders: HashMap::new(),
+            balances: HashMap::new(),
+            tokens: Tokens(Vec::new()),
+            pending_solution: PendingSolution::AccumulatingTrades(Vec::new()),
```

Comment:
nit: would it make sense to implement default on PendingSolution instead and derive it here?